File: docs/garden/src/zig-comptime/macros.c
#include <stdio.h>

/* Препроцессор подставляет токены, ничего не зная о выражениях и именах. */
// --8<-- [start:defs]
#define SQUARE(x) x * x
#define MAX(a, b) ((a) > (b) ? (a) : (b))
#define PRINT_TWICE(e)                                                         \
    do {                                                                       \
        int i;                                                                 \
        for (i = 0; i < 2; i++)                                                \
            printf("  %d\n", (e));                                             \
    } while (0)
// --8<-- [end:defs]

int main(void) {
    // --8<-- [start:use]
    /* 1. Приоритет: раскрывается в 1 + 2 * 1 + 2. */
    printf("SQUARE(1 + 2) = %d\n", SQUARE(1 + 2));

    /* 2. Двойное вычисление: i++ выполняется дважды. */
    int i = 3, j = 2;
    int m = MAX(i++, j);
    printf("MAX(i++, j) = %d, i = %d\n", m, i);

    /* 3. Захват имени: i в аргументе связывается с i внутри макроса. */
    i = 7;
    printf("PRINT_TWICE(i * 10):\n");
    PRINT_TWICE(i * 10);
    // --8<-- [end:use]
    return 0;
}

File: docs/garden/src/zig-comptime/macros.rs
// --8<-- [start:code]
// macro_rules! работает с деревьями токенов, а фрагмент $x:expr —
// уже разобранное выражение: скобки вокруг него не нужны.
macro_rules! square {
    ($x:expr) => {
        $x * $x
    };
}

// Аргумент вычисляется один раз, потому что макрос связывает его с let.
macro_rules! max {
    ($a:expr, $b:expr) => {{
        let a = $a;
        let b = $b;
        if a > b { a } else { b }
    }};
}

// Гигиена: i внутри макроса и i в месте вызова — разные переменные.
macro_rules! print_twice {
    ($e:expr) => {
        for i in 0..2 {
            let _ = i;
            println!("  {}", $e);
        }
    };
}

// const fn — отдельный, более узкий режим того же языка.
const fn fib(n: u32) -> u32 {
    if n < 2 { n } else { fib(n - 1) + fib(n - 2) }
}
const FIB10: u32 = fib(10);

fn main() {
    println!("square!(1 + 2) = {}", square!(1 + 2));

    let mut i = 3;
    let m = max!({ i += 1; i - 1 }, 2); // «i++» в стиле Rust
    println!("max!(i++, 2) = {m}, i = {i}");

    let i = 7;
    println!("print_twice!(i * 10):");
    print_twice!(i * 10);

    let buf = [0u8; fib(6) as usize];
    println!("FIB10 = {FIB10}, buf.len() = {}", buf.len());
}
// --8<-- [end:code]
